tactic/ext.lean

```diff
@@ -1,23 +1,142 @@
 
-import tactic.basic
-
+import tactic.basic core data.sum tactic.rcases
 universes u₁ u₂
 
+open interactive interactive.types
+open lean.parser nat tactic
+
+meta def get_ext_subject : expr → tactic name
+| (expr.pi n bi d b) :=
+  do v  ← mk_local' n bi d,
+     b' ← whnf $ b.instantiate_var v,
+     get_ext_subject b'
+| (expr.app _ e) :=
+  do t ← infer_type e,
+     if t.is_app ∨ t.is_constant then
+       pure $ t.get_app_fn.const_name
+     else if t.is_pi then
+       pure name.anonymous
+     else failed
+| _ := failed
+
+open native
+
+@[reducible] def ext_param_type := option name ⊕ option name
+
+meta def opt_minus : lean.parser (option name → ext_param_type) :=
+sum.inl <$ tk "-" <|> pure sum.inr
+
+meta def ext_param :=
+opt_minus <*> ( name.anonymous <$ brackets "(" ")" (tk "→" <|> tk "->") <|>
+                none <$  tk "*" <|>
+                some <$> ident )
+
+meta def saturate_fun : name → tactic expr
+| name.anonymous :=
+do v₀ ← mk_mvar,
+   v₁ ← mk_mvar,
+   return $ v₀.imp v₁
+| n :=
+do e ← resolve_constant n >>= mk_const,
+   a ← get_arity e,
+   e.mk_app <$> (list.iota a).mmap (λ _, mk_mvar)
+
+meta def equiv_type_constr (n n' : name) : tactic unit :=
+do e  ← saturate_fun n,
+   e' ← saturate_fun n',
+   unify e e' <|> fail format!"{n} and {n'} are not definitionally equal types"
+
+run_cmd equiv_type_constr `list `list
+
 /--
  Tag lemmas of the form:
 
  ```
+ @[extensionality]
+ lemma my_collection.ext (a b : my_collection)
+   (h : ∀ x, a.lookup x = b.lookup y) :
+   a = b := ...
+ ```
+
+ The attribute indexes extensionality lemma using the type of the
+ objects (i.e. `my_collection`) which it gets from the statement of
+ the lemma.  In some cases, the same lemma can be used to state the
+ extensionality of multiple types that are definitionally equivalent.
+
+ ```
+ attribute [extensionality [(→),thunk,stream]] funext
+ ```
+
+ Those parameters are cumulative. The following are equivalent:
+
+ ```
+ attribute [extensionality [(→),thunk]] funext
+ attribute [extensionality [stream]] funext
+ ```
+ and
+ ```
+ attribute [extensionality [(→),thunk,stream]] funext
+ ```
+
+ One removes type names from the list for one lemma with:
+ ```
+ attribute [extensionality [-stream,-thunk]] funext
+  ```
+
+ Finally, the following:
+
+ ```
+ @[extensionality]
+ lemma my_collection.ext (a b : my_collection)
+   (h : ∀ x, a.lookup x = b.lookup y) :
+   a = b := ...
+ ```
+
+ is equivalent to
+
+ ```
+ @[extensionality]
+ lemma my_collection.ext (a b : my_collection)
+   (h : ∀ x, a.lookup x = b.lookup y) :
+   a = b := ...
+ ```
```

Comment:
These two examples are the same?

Reply:
@cipher1024 the `*` is still missing here.